README.md
<h1 align="center">qemu2deb</h1>

<p align="center">
    <a href="https://github.com/Itai-Nelken/qemu2deb/" alt="made with bash badge">
        <img src="https://img.shields.io/badge/Made%20with-bash-green?style=for-the-badge&logo=GNU%20bash" /></a>
    <a href="https://github.com/Itai-Nelken/qemu2deb/blob/main/LICENSE" alt="license badge">
        <img src="https://img.shields.io/github/license/Itai-Nelken/qemu2deb?style=for-the-badge" /></a>
    <a href="https://github.com/Itai-Nelken/qemu2deb/releases/latest" alt="latest release badge">
        <img src="https://img.shields.io/github/v/release/Itai-Nelken/qemu2deb?include_prereleases&style=for-the-badge" /></a>
    <a href="https://github.com/Itai-Nelken/qemu2deb/releases/" alt="Downloads">
        <img src="https://img.shields.io/github/downloads/Itai-Nelken/qemu2deb/total?style=for-the-badge" /></a>
    <!--
    <a href="https://github.com/Itai-Nelken/qemu2deb/blob/main/qemu2deb.sh" alt="script size badge">
        <img src="https://img.shields.io/github/size/Itai-Nelken/qemu2deb/qemu2deb.sh?style=for-the-badge" /></a>
    <a href="https://github.com/Itai-Nelken/qemu2deb/" alt="fun to use?">
        <img src="https://img.shields.io/badge/fun%20to%20use%3F-YES-blue?style=for-the-badge&logo=qemu" /></a>
    -->
</p>

<p align="center">Compiling/packaging/installing QEMU on the Raspberry Pi and other Linux computers made easier than ever.</p>

## Installation and usage:
<!--
1) Download the script: you can get the latest release [here](https://raw.githubusercontent.com/Itai-Nelken/qemu2deb-RPi/0.3.1/qemu2deb.sh) the newest "bleeding edge" stable version [here](https://raw.githubusercontent.com/Itai-Nelken/qemu2deb-RPi/main/qemu2deb.sh), and the Beta version [here](https://raw.githubusercontent.com/Itai-Nelken/qemu2deb-RPi/Dev/qemu2deb.sh) (the Beta version isn't always available). you can use `wget` to download the script or simply copy-and-paste the contents to a file called `qemu2deb.sh`.
2) give it executable permissions: `sudo chmod +x qemu2deb.sh`
-->
1) Download the deb file from the latest release [here](https://github.com/Itai-Nelken/qemu2deb-RPi/releases/latest).
2) Install it using this command: `sudo dpkg -i qemu2deb*.deb` or using your favorite graphical app/command.
3) to run the qemu2deb script, type the following in terminal (works from any location): `qemu2deb`<br>
**For a full walkthrough, read [USAGE.md](https://github.com/Itai-Nelken/qemu2deb/blob/main/USAGE.md#detailed-usage-instructions)**<br>
NOTE: don't run the script as root! if you do, the script will warn you and exit.<br>
if you have passwordless root access **disabled**, you will have to enter your password at least once while running the script.

## How it works
First the script asks you where is your working directory, then if you tell it QEMU isn't compiled and installed yet, it will compile and install it.
after it has the QEMU directory and your working directory and QEMU is installed on your system, it will copy all the QEMU files to a temporary folder, create the DEBIAN/control file and package the temporary folder to a deb using `dpkg-deb`.
after packaging is complete, the script will clean up after itself.

## [Changelog](CHANGELOG.md)

## compatibility list:
### OS's
- [x] RPiOS and RPiOS based OS's for the pi (like TwisterOS).
- [x] Debian and Debian based OS's (including Ubuntu and Ubuntu based OS's of course).
- [ ] Android with [Termux](https://termux.com/).
### Architectures
- [x] all common architectures (armhf, arm64, x86/i386 and x86_64/x64/amd64).

### have a fix, suggestion or a bug to report? feel free to [open a issue](https://github.com/Itai-Nelken/qemu2deb-RPi/issues/new/choose) or a Pull request!


## FAQ:
### Q = question<br>A = answer

**Q:**
>why did you create this script?

**A:**
>because compiling QEMU on the Raspberry Pi takes over an hour and the version in the repos is really old, and I wanted to have a easy and fast way to install QEMU because I switch OS's a lot.
first I tried using checkinstall, but it didn't work, so after trying a lot of fixes and ways I decided to create the deb manually.
I tracked down all the qemu files on my system, put them in a folder, created the DEBIAN/control file, and pakaged it. to my surprise it worked, but I got a error that the bios wasn't found. after searching a bit I found the issue: broken links. I found the files the links pointed to and copied them over to the folder, then packaged it again.
this time I messed up the two debs, so I deleted both debs, and packaged one again. this time it worked!
so I started to write this script for my own use. once I saw how well it was working and how much time it was saving me, I decided to make a repo for it so other people can use it.
I hope you enjoy using this script and find it as useful as I find it, it took me a week to get it working correctly, and almost a whole year of trying different ways to package QEMU, and I don't want anyone else to have to go through this.


CHANGELOG.md
# QEMU2DEB Changelog

**DD/MM/YY**

### 23/01/2021
I was struggling to package QEMU to a deb using all the tools I knew of.

### 24/01/2021
I found a easy way to package QEMU to a deb, and wrote this script.

### 25/01/2021
Initial upload, lots of bug fixes and quality of life improvements, first 2 releases (0.1.0, 0.1.1)!
at this point the script is fully functional but lacking some features I would like to add to it (like more control over the DEBIAN/control file).

### 30/1/2021
update release to 0.1.2: lots of minor changes and bug fixes.
added a lot of the conflicting packages to the deb's control file, understood why [`apt` thinks its downgrading](https://unix.stackexchange.com/questions/631805/how-to-hold-package-from-updating-with-posttest-script-in-deb) and getting ready to add a fix for the next major release.
added root detection - that means that the script won't let you run it as root.

### 31/1/2021
update release to 0.2! - fixed apt thinking its downgrading.
slight update to the `install-deb` function. releases are not marked as pre-releases anymore!
opened 'Dev' branch for testing without breaking what is already working.
added 'error' function to report errors in a easy to understand format and exit.
getting ready to merge the Dev branch to main and bump version to 0.3.0!

### 1/2/2021
Update release to 0.3.0!

### 9/2/2021
update release to 0.3.1 - added `priority` and `section` to control file.

### 14/2/2021
update release to 0.3.2 - add shorter versions of flags (`-h` = `--help`, `-v` = `--version`).
the script now checks if you are on a armhf OS, else it will warn you and exit. to disable checking, use the `--no-check-arch` flag (also added).
qemu2deb is now available as a deb! (I also added a deb to some of the old releases).

### 15/2/2021
Update to 0.3.3 - The help printed when using the `--help` flag is more extensive and now includes instructions on how to use if you installed qemu2deb from the deb.

### 19/2/2021
Released 0.4.0: GIANT UPDATES!
  1) better architecture detection: now the script checks if you are on armhf or arm64, if yes it checks if you are on a 32bit or 64bit OS.
  2) replace all `clear` commands with `clear -x` - that means its way easier to debug.
  3) updated the `install-deb` and `clean-up` function, now they work way better.
  4) replaced the `install-depends` function with `apt-install` and made a variable with all the dependencies, so now I can run: `apt-install $DEPENDS` to install all dependencies.
  5) replaced `make` with `ninja` for compiling qemu, its almost twice as fast.
  6) updated the`make-deb` function to make it work again. now the files it copies make more sense and will probably allow this script to work on arm64 and x86 OS's!
and probably a bit more I forgot to document.

### 21/2/2021
control file uses the "$ARCH" variable in the descritpion field.
Replaced the `apt-install` function with `pkg-manage` function, this new function allows me to install and uninstall packages. this function brings me to the next change: the script now asks you if you wan't to uninstall the dependencies when cleaning up after itself though it isn't recommended to do it yet because it will uninstall all the packages even if you had them installed before, that brings me to a change I'm working on: making the `pkg-manage` function remember wich packages it installed, and then ask to uninstall only the ones that it installed.
added support for x86 and x86_64!, arch checking is much simpler now as well.

### 22/2/2021
added Raspberry Pi detection that is used to decide what is the summary and description of the deb.
some fixes for x86_64 including changing the output of x86_64 to amd64 because its the correct way to put in the deb control file.
give the finished deb '606' permissions.

### 27/2/2021
made the `pkg-manage` function check what packages are installed from the list provided to it, and save the ones that aren't installed to the 'TOINSTALL' variable and install those.
also added to it a 'check' option to check what packages are installed and save them to the 'TOINSTALL' variable.
improve all the places where a question is asked, fixed a few typos, some quality of life improvements.

### 28/2/2021
impelement ctrl+c trapping: if you press ctrl+c the script will clean up after itself.

### 2/3/2021
make the script not copy any bash completion files as they are in the 'bash-completion' package, and including them will cause the deb to fail to install.
add 'bash-completion' as a recommended package to the deb's control file.
a few quality of life improvements.

### 4/3/2021
add clean option to the `pkg-manage` function, it runs apt clean, autoremove, and autoclean. it is used when uninstalling the dependencies after building QEMU.

### 5/3/2021
released version 0.6.0!

### 14/3/2021
Removed `/usr/local/` from the deb. all files copied from there are put in `/usr/` instead. 

### ~15/3/2021-24/4/2021
**GIANT CHANGES AND IMPROVEMENTS!**
1) The install-deb function was depracated and removed, the clean-up function does the work now.
2) Updated the clean-up function.
3) Fixed some issues with the "$QBUILD" variable (where to compile QEMU).
4) Re-wrote the error function, now it foes the work of the warning and warning-sleep functions.
5) Improvements to the make-deb function and renamed it to cp-files, added error handling.
6) Added variables holding the escape character "codes" used to format text.
7) Completely re-wrote the ctrl_c function! now it know in what part of the script it was called, and cleans up according to that. 
8) Added the "$CALLCOMMAND" variable that lets me set the name of the script to use in the help function.
9) Change to 'case' instead of 'if-then-elif-else' statment for the flags.
10) Added the 'maintainer=<string>' flag, so you can set the maintainer name when running the script!.
11) If user selected to compile QEMU, ask where.
12) bumped version to 0.7.0
13) print errors to `stderr`

qemu2deb-dev.sh
#!/bin/bash

#check that script isn't being run as root.
if [ "$EUID" = 0 ]; then
  echo "You cannot run this script as root!"
  exit 1
fi

#variables
CORES="`nproc`"
#determine if host system is 64 bit arm64 or 32 bit armhf
if [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 64)" ];then
  SARCH=64
elif [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 32)" ];then
  SARCH=32
else
  echo -e "$(tput setaf 1)$(tput bold)Can't detect OS architecture! something is very wrong!$(tput sgr 0)"
  exit
fi

#script version variable
APPVER="0.3.0"
#remove Beta variable for final release
BETA="beta 1"

#functions
function intro() {
    echo "this script doesn't do anything and exists only for my testing."
    echo -e "
    #################################################
    #  QEMU2DEB $APPVER $BETA by Itai-Nelken | 2021  #
    #-----------------------------------------------#
    #     compile/package/install QEMU              #
    #################################################
    "
}

function help() {
    echo "$(tput bold)$(tput setaf 6)usage:$(tput sgr 0)"
    echo "./qemu2deb.sh [flags]"
    echo " "
    echo "$(tput setaf 6)available flags:$(tput sgr 0)"
    echo "--version  -  display version and exit."
    echo "--help  -  display this help."
}

function install-deb() {
    read -p "do you want to install the DEB (y/n)?" choice
    case "$choice" in 
        y|Y ) CONTINUE=1 ;;
        n|N ) CONTINUE=0 ;;
        * ) echo "invalid" ;;
    esac

if [[ "$CONTINUE" == 1 ]]; then
    echo "cd $DIRECTORY
    sudo apt -f -y install ./qemu-$QVER-$ARCH.deb"
elif [[ "$CONTINUE" == 0 ]]; then
    clear
fi
}

function clean-up() {
echo -e "$(tput setaf 3)$(tput bold)cleaning up...$(tput sgr 0)"
sleep 0.3
read -p "do you want to delete the qemu build folder (y/n)?" choice
case "$choice" in 
  y|Y ) CONTINUE=1 ;;
  n|N ) CONTINUE=0 ;;
  * ) echo "invalid" ;;
esac

if [[ "$CONTINUE" == 1 ]]; then
    echo "cd $QBUILD
    cd ..
    sudo rm -r qemu"
elif [[ "$CONTINUE" == 0 ]]; then
    echo "won't remove $QBUILD"
fi

read -p "do you want to delete the unpacked DEB (y/n)?" choice
case "$choice" in 
  y|Y ) CONTINUE=1 ;;
  n|N ) CONTINUE=0 ;;
  * ) echo "invalid" ;;
esac

if [[ "$CONTINUE" == 1 ]]; then
    echo "sudo rm -r qemu-$QVER-$ARCH"
elif [[ "$CONTINUE" == 0 ]]; then
    echo "won't remove unpacked DEB"
fi
}

function install-depends() {
    echo "sudo apt install -y build-essential ninja-build libepoxy-dev libdrm-dev libgbm-dev libx11-dev libvirglrenderer-dev libpulse-dev libsdl2-dev git libglib2.0-dev libfdt-dev libpixman-1-dev zlib1g-dev libepoxy-dev libdrm-dev libgbm-dev libx11-dev libvirglrenderer-dev libpulse-dev libsdl2-dev"
}

function compile-qemu() {
    echo '
    cd $DIRECTORY'
    echo "cloning QEMU git repo..."
    echo 'git clone https://git.qemu.org/git/qemu.git
    cd qemu
    git submodule init
    git submodule update --recursive
    echo "running ./configure..."
    ./configure --enable-sdl  --enable-opengl --enable-virglrenderer --enable-system --enable-modules --audio-drv-list=pa --enable-kvm
    echo "compiling QEMU..."
    make -j$CORES
    sudo make install'
}

function make-deb() {
    #get QEMU version
    QVER="`qemu-system-ppc --version | grep version | cut -c23-28`"
    #get arch
    if [[ "$SARCH" == 64 ]]; then
        ARCH=arm64
    elif [[ "$SARCH" == 32 ]]; then
     ARCH=armhf
    fi
    #get all files inside a folder before building deb
    echo "
    clear
    echo "copying files..."
    echo -ne '(0%)[#                         ](100%)\r'
    cd $DIRECTORY
    mkdir qemu-$QVER-$ARCH
    cd qemu-$QVER-$ARCH
    mkdir -p usr/bin/
    cd usr
    echo -ne '(0%)[###                       ](100%)\r'
    sudo cp /usr/bin/qemu* $DIRECTORY/qemu-$QVER-$ARCH/usr/bin/
    mkdir -p lib/arm-linux-gnueabihf/qemu
    sudo cp -r /usr/lib/arm-linux-gnueabihf/qemu $DIRECTORY/qemu-$QVER-$ARCH/usr/lib/arm-linux-gnueabihf/
    sudo cp -r /usr/lib/qemu/ $DIRECTORY/qemu-$QVER-$ARCH/usr/lib/
    echo -ne '(0%)[########                  ](100%)\r'
    mkdir -p local/bin/
    mkdir -p local/lib/qemu
    mkdir -p local/share/qemu
    sudo cp /usr/local/bin/qemu* $DIRECTORY/qemu-$QVER-$ARCH/usr/local/bin/
    sleep 0.1
    echo -ne '(0%)[##########                ](100%)\r'
    sudo cp -r /usr/local/lib/qemu $DIRECTORY/qemu-$QVER-$ARCH/usr/local/lib/
    sleep 0.1
    echo -ne '(0%)[#############             ](100%)\r'
    sudo cp -r /usr/local/share/qemu/ $DIRECTORY/qemu-$QVER-$ARCH/usr/local/share/
    mkdir -p share/man/man1
    sudo cp /usr/share/man/man1/qemu*.1.gz $DIRECTORY/qemu-$QVER-$ARCH/usr/share/man/man1
    sleep 0.1
    echo -ne '(0%)[#################         ](100%)\r'
    mkdir -p share/openbios
    mkdir -p share/openhackware
    mkdir -p share/ovmf
    mkdir -p share/qemu
    mkdir -p share/slof
    sudo cp -r /usr/share/openbios/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    sudo cp -r /usr/share/openhackware/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    sleep 0.1
    echo -ne '(0%)[####################     ](100%)\r'
    sudo cp -r /usr/share/ovmf/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    sudo cp -r /usr/share/qemu/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    sudo cp -r /usr/share/slof/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    cd ..
    sleep 0.1
    echo -ne '(0%)[#########################](100%)\r'"
    sleep 1
}


##########flags##########
if  [[ $1 = "--version" ]]; then
    intro
    exit
elif [[ $1 = "--help" ]]; then
    help
    exit
fi



##########The part where things actually start to happen##########

#clear the screen
clear
#run the "intro" function
intro
#print a blank line
echo ' '
#ask for directory path, if doesn't exist ask again. if exists exit loop.
while true;
    do
            read -p "Enter full path to directory where you want to make the deb:" DIRECTORY
            if [ ! -d $DIRECTORY ]; then
                    echo "directory does not exist, please try again"

            else
                    echo "$(tput bold)qemu will be built and packaged here: $DIRECTORY$(tput sgr 0)"
                    break
            fi
done

#sleep 2 seconds and clear the screen
sleep 2
echo " "
#ask if you already compiled QEMU, if yes enter full path (same as other loop), if you press s, the loop exits.
while true;
    do
            read -p "If you already compiled and installed QEMU (with sudo make install), enter the path to its folder. otherwise press s:" QBUILD
            if [[ "$QBUILD" == s ]]; then
                echo "QEMU will be compiled..."
                QBUILDV=1
                break
            fi
            if [ ! -d $QBUILD ]; then
                    echo "directory does not exist, please try again"

            else
                    echo "$(tput bold)qemu is already built here: $QBUILD$(tput sgr 0)"
                    QBUILDV=0
                    break
            fi
done

#wait 1.5 seconds and clear the screen
sleep 1.5
clear

#if QEMU needs to be compiled, do so
if [[ "$QBUILDV" == 1 ]]; then
    echo -e "$(tput setaf 6)$(tput bold)QEMU will now be compiled, this will take over a hour and consume all CPU.$(tput sgr 0)"
    echo -e "$(tput setaf 6)$(tput bold)cooling is recommended.$(tput sgr 0)"
    read -p "Press [ENTER] to continue"
    install-depends
    compile-qemu
elif [[ "$QBUILDV" == 0 ]]; then
    read -p "do you want to install QEMU (run 'sudo make install') (y/n)?" choice
    case "$choice" in 
      y|Y ) CONTINUE=1 ;;
      n|N ) CONTINUE=0 ;;
      * ) echo "invalid" ;;
    esac
    if [[ "$CONTINUE" == 1 ]]; then
        echo "cd $QBUILD
        sudo make install"
    elif [[ "$CONTINUE" == 0 ]]; then
        if [ ! command -v qemu-img &>/dev/null ];then
            echo "QEMU is installed..."
        else
            echo "$(tput setaf 1)QEMU isn't installed! can't continue$(tput bold)$(tput sgr 0)"
            exit 1
        fi
    fi
fi

sleep 3
#clear the screen again
clear
#print the summary so far and ask to continue
printf "$(tput bold)\\e[3;4;37mSummary:\\n\\e[0m$(tput sgr 0)"
echo "the DEB will be built here: $DIRECTORY"
if [[ "$QBUILDV" == 1 ]]; then
    echo "QEMU was compiled here: $DIRECTORY"
elif [[ "$QBUILDV" == 0 ]]; then
    echo "QEMU is already compiled here: $QBUILD"
fi
read -p "Press [ENTER] to continue or [CTRL+C] to cancel"


#start making the deb folder (unpacked deb)
echo -e "$(tput setaf 6)$(tput bold)QEMU will now be packaged into a DEB, this will take a few minutes and consume all CPU.$(tput sgr 0)"
echo -e "$(tput setaf 6) $(tput bold)cooling is recommended. $(tput sgr 0)"
read -p "Press [ENTER] to continue"
#copy all files using the 'make-deb' function
make-deb
echo "creating DEBIAN folder..."
echo "
mkdir DEBIAN
cd DEBIAN"
sleep 2
echo "clear"
echo "creating control file..."
#ask for maintainer info
echo -e "$(tput setaf 3)$(tput bold)enter maintainer info:$(tput sgr 0)"
read MAINTAINER
clear

#create DEBIAN/control
echo "
Maintainer: $MAINTAINER 
Summary: QEMU $QVER armhf for the raspberry pi built using qemu2deb
Name: qemu 
Description: QEMU $QVER $ARCH built using QEMU2DEB for arm devices.
Version: 1:$QVER 
Release: 1 
License: GPL 
Architecture: $ARCH 
Provides:qemu
Conflicts:qemu-utils, qemu-system-common, qemu-system-gui, qemu-system-ppc, qemu-block-extra, qemu-guest-agent, qemu-kvm, qemu-system-arm, qemu-system-common, qemu-system-mips, qemu-system-misc, qemu-system-sparc, qemu-system-x86, qemu-system, qemu-user-binfmt, qemu-user-static, qemu-user, qemu, openbios-sparc, openbios-ppc, seabios
Package: qemu > control
#give it the necessary permissions
sudo chmod 775 control
cd ..
cd ..
#build the DEB
sudo dpkg-deb --build qemu-$QVER-$ARCH/"

echo -e "$(tput setaf 3)$(tput bold)DONE...$(tput sgr 0)"
echo "qemu deb will be in $DIRECTORY/qemu-$QVER-$ARCH.deb"
read -p "Press [ENTER] to continue"
clear

#ask to install the deb snd then clean up
install-deb
clean-up

echo -e "$(tput setaf 2)$(tput bold)DONE...$(tput sgr 0)"
echo "exiting in 10 seconds..."
sleep 10
exit


qemu2dev-dev.sh
#!/bin/bash

#check that script isn't being run as root.
if [ "$EUID" = 0 ]; then
  echo "You cannot run this script as root!"
  exit 1
fi

#variables
CORES="`nproc`"
#determine if host system is 64 bit arm64 or 32 bit armhf
if [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 64)" ];then
  SARCH=64
elif [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 32)" ];then
  SARCH=32
else
  echo -e "$(tput setaf 1)$(tput bold)Can't detect OS architecture! something is very wrong!$(tput sgr 0)"
  exit
fi

#script version variable
APPVER="0.3.0"
#remove Beta variable for final release
BETA="beta 1"

#functions
function intro() {
    echo "this script doesn't do anything and exists only for my testing."
    echo -e "
    #################################################
    #  QEMU2DEB $APPVER $BETA by Itai-Nelken | 2021  #
    #-----------------------------------------------#
    #     compile/package/install QEMU              #
    #################################################
    "
}

function help() {
    echo "$(tput bold)$(tput setaf 6)usage:$(tput sgr 0)"
    echo "./qemu2deb.sh [flags]"
    echo " "
    echo "$(tput setaf 6)available flags:$(tput sgr 0)"
    echo "--version  -  display version and exit."
    echo "--help  -  display this help."
}

function install-deb() {
    read -p "do you want to install the DEB (y/n)?" choice
    case "$choice" in 
        y|Y ) CONTINUE=1 ;;
        n|N ) CONTINUE=0 ;;
        * ) echo "invalid" ;;
    esac

if [[ "$CONTINUE" == 1 ]]; then
    echo "cd $DIRECTORY
    sudo apt -f -y install ./qemu-$QVER-$ARCH.deb"
elif [[ "$CONTINUE" == 0 ]]; then
    clear
fi
}

function clean-up() {
echo -e "$(tput setaf 3)$(tput bold)cleaning up...$(tput sgr 0)"
sleep 0.3
read -p "do you want to delete the qemu build folder (y/n)?" choice
case "$choice" in 
  y|Y ) CONTINUE=1 ;;
  n|N ) CONTINUE=0 ;;
  * ) echo "invalid" ;;
esac

if [[ "$CONTINUE" == 1 ]]; then
    echo "cd $QBUILD
    cd ..
    sudo rm -r qemu"
elif [[ "$CONTINUE" == 0 ]]; then
    echo "won't remove $QBUILD"
fi

read -p "do you want to delete the unpacked DEB (y/n)?" choice
case "$choice" in 
  y|Y ) CONTINUE=1 ;;
  n|N ) CONTINUE=0 ;;
  * ) echo "invalid" ;;
esac

if [[ "$CONTINUE" == 1 ]]; then
    echo "sudo rm -r qemu-$QVER-$ARCH"
elif [[ "$CONTINUE" == 0 ]]; then
    echo "won't remove unpacked DEB"
fi
}

function install-depends() {
    echo "sudo apt install -y build-essential ninja-build libepoxy-dev libdrm-dev libgbm-dev libx11-dev libvirglrenderer-dev libpulse-dev libsdl2-dev git libglib2.0-dev libfdt-dev libpixman-1-dev zlib1g-dev libepoxy-dev libdrm-dev libgbm-dev libx11-dev libvirglrenderer-dev libpulse-dev libsdl2-dev"
}

function compile-qemu() {
    echo '
    cd $DIRECTORY'
    echo "cloning QEMU git repo..."
    echo 'git clone https://git.qemu.org/git/qemu.git
    cd qemu
    git submodule init
    git submodule update --recursive
    echo "running ./configure..."
    ./configure --enable-sdl  --enable-opengl --enable-virglrenderer --enable-system --enable-modules --audio-drv-list=pa --enable-kvm
    echo "compiling QEMU..."
    make -j$CORES
    sudo make install'
}

function make-deb() {
    #get QEMU version
    QVER="`qemu-system-ppc --version | grep version | cut -c23-28`"
    #get arch
    if [[ "$SARCH" == 64 ]]; then
        ARCH=arm64
    elif [[ "$SARCH" == 32 ]]; then
     ARCH=armhf
    fi
    #get all files inside a folder before building deb
    echo "
    clear
    echo "copying files..."
    echo -ne '(0%)[#                         ](100%)\r'
    cd $DIRECTORY
    mkdir qemu-$QVER-$ARCH
    cd qemu-$QVER-$ARCH
    mkdir -p usr/bin/
    cd usr
    echo -ne '(0%)[###                       ](100%)\r'
    sudo cp /usr/bin/qemu* $DIRECTORY/qemu-$QVER-$ARCH/usr/bin/
    mkdir -p lib/arm-linux-gnueabihf/qemu
    sudo cp -r /usr/lib/arm-linux-gnueabihf/qemu $DIRECTORY/qemu-$QVER-$ARCH/usr/lib/arm-linux-gnueabihf/
    sudo cp -r /usr/lib/qemu/ $DIRECTORY/qemu-$QVER-$ARCH/usr/lib/
    echo -ne '(0%)[########                  ](100%)\r'
    mkdir -p local/bin/
    mkdir -p local/lib/qemu
    mkdir -p local/share/qemu
    sudo cp /usr/local/bin/qemu* $DIRECTORY/qemu-$QVER-$ARCH/usr/local/bin/
    sleep 0.1
    echo -ne '(0%)[##########                ](100%)\r'
    sudo cp -r /usr/local/lib/qemu $DIRECTORY/qemu-$QVER-$ARCH/usr/local/lib/
    sleep 0.1
    echo -ne '(0%)[#############             ](100%)\r'
    sudo cp -r /usr/local/share/qemu/ $DIRECTORY/qemu-$QVER-$ARCH/usr/local/share/
    mkdir -p share/man/man1
    sudo cp /usr/share/man/man1/qemu*.1.gz $DIRECTORY/qemu-$QVER-$ARCH/usr/share/man/man1
    sleep 0.1
    echo -ne '(0%)[#################         ](100%)\r'
    mkdir -p share/openbios
    mkdir -p share/openhackware
    mkdir -p share/ovmf
    mkdir -p share/qemu
    mkdir -p share/slof
    sudo cp -r /usr/share/openbios/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    sudo cp -r /usr/share/openhackware/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    sleep 0.1
    echo -ne '(0%)[####################     ](100%)\r'
    sudo cp -r /usr/share/ovmf/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    sudo cp -r /usr/share/qemu/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    sudo cp -r /usr/share/slof/ $DIRECTORY/qemu-$QVER-$ARCH/usr/share/
    cd ..
    sleep 0.1
    echo -ne '(0%)[#########################](100%)\r'"
    sleep 1
}


##########flags##########
if  [[ $1 = "--version" ]]; then
    intro
    exit
elif [[ $1 = "--help" ]]; then
    help
    exit
fi



##########The part where things actually start to happen##########

#clear the screen
clear
#run the "intro" function
intro
#print a blank line
echo ' '
#ask for directory path, if doesn't exist ask again. if exists exit loop.
while true;
    do
            read -p "Enter full path to directory where you want to make the deb:" DIRECTORY
            if [ ! -d $DIRECTORY ]; then
                    echo "directory does not exist, please try again"

            else
                    echo "$(tput bold)qemu will be built and packaged here: $DIRECTORY$(tput sgr 0)"
                    break
            fi
done

#sleep 2 seconds and clear the screen
sleep 2
echo " "
#ask if you already compiled QEMU, if yes enter full path (same as other loop), if you press s, the loop exits.
while true;
    do
            read -p "If you already compiled and installed QEMU (with sudo make install), enter the path to its folder. otherwise press s:" QBUILD
            if [[ "$QBUILD" == s ]]; then
                echo "QEMU will be compiled..."
                QBUILDV=1
                break
            fi
            if [ ! -d $QBUILD ]; then
                    echo "directory does not exist, please try again"

            else
                    echo "$(tput bold)qemu is already built here: $QBUILD$(tput sgr 0)"
                    QBUILDV=0
                    break
            fi
done

#wait 1.5 seconds and clear the screen
sleep 1.5
clear

#if QEMU needs to be compiled, do so
if [[ "$QBUILDV" == 1 ]]; then
    echo -e "$(tput setaf 6)$(tput bold)QEMU will now be compiled, this will take over a hour and consume all CPU.$(tput sgr 0)"
    echo -e "$(tput setaf 6)$(tput bold)cooling is recommended.$(tput sgr 0)"
    read -p "Press [ENTER] to continue"
    install-depends
    compile-qemu
elif [[ "$QBUILDV" == 0 ]]; then
    read -p "do you want to install QEMU (run 'sudo make install') (y/n)?" choice
    case "$choice" in 
      y|Y ) CONTINUE=1 ;;
      n|N ) CONTINUE=0 ;;
      * ) echo "invalid" ;;
    esac
    if [[ "$CONTINUE" == 1 ]]; then
        echo "cd $QBUILD
        sudo make install"
    elif [[ "$CONTINUE" == 0 ]]; then
        if [ ! command -v qemu-img &>/dev/null ];then
            echo "$(tput setaf 1)QEMU isn't installed! can't continue$(tput bold)$(tput sgr 0)"
            exit 1
        else
            echo "QEMU is installed..."
        fi
    fi
fi

sleep 3
#clear the screen again
clear
#print the summary so far and ask to continue
printf "$(tput bold)\\e[3;4;37mSummary:\\n\\e[0m$(tput sgr 0)"
echo "the DEB will be built here: $DIRECTORY"
if [[ "$QBUILDV" == 1 ]]; then
    echo "QEMU was compiled here: $DIRECTORY"
elif [[ "$QBUILDV" == 0 ]]; then
    echo "QEMU is already compiled here: $QBUILD"
fi
read -p "Press [ENTER] to continue or [CTRL+C] to cancel"


#start making the deb folder (unpacked deb)
echo -e "$(tput setaf 6)$(tput bold)QEMU will now be packaged into a DEB, this will take a few minutes and consume all CPU.$(tput sgr 0)"
echo -e "$(tput setaf 6) $(tput bold)cooling is recommended. $(tput sgr 0)"
read -p "Press [ENTER] to continue"
#copy all files using the 'make-deb' function
make-deb
echo "creating DEBIAN folder..."
echo "
mkdir DEBIAN
cd DEBIAN"
sleep 2
echo "clear"
echo "creating control file..."
#ask for maintainer info
echo -e "$(tput setaf 3)$(tput bold)enter maintainer info:$(tput sgr 0)"
read MAINTAINER
clear

#create DEBIAN/control
echo "
Maintainer: $MAINTAINER 
Summary: QEMU $QVER armhf for the raspberry pi built using qemu2deb
Name: qemu 
Description: QEMU $QVER $ARCH built using QEMU2DEB for arm devices.
Version: 1:$QVER 
Release: 1 
License: GPL 
Architecture: $ARCH 
Provides:qemu
Conflicts:qemu-utils, qemu-system-common, qemu-system-gui, qemu-system-ppc, qemu-block-extra, qemu-guest-agent, qemu-kvm, qemu-system-arm, qemu-system-common, qemu-system-mips, qemu-system-misc, qemu-system-sparc, qemu-system-x86, qemu-system, qemu-user-binfmt, qemu-user-static, qemu-user, qemu, openbios-sparc, openbios-ppc, seabios
Package: qemu > control
#give it the necessary permissions
sudo chmod 775 control
cd ..
cd ..
#build the DEB
sudo dpkg-deb --build qemu-$QVER-$ARCH/"

echo -e "$(tput setaf 3)$(tput bold)DONE...$(tput sgr 0)"
echo "qemu deb will be in $DIRECTORY/qemu-$QVER-$ARCH.deb"
read -p "Press [ENTER] to continue"
clear

#ask to install the deb snd then clean up
install-deb
clean-up

echo -e "$(tput setaf 2)$(tput bold)DONE...$(tput sgr 0)"
echo "exiting in 10 seconds..."
sleep 10
exit
